AsyncDisplayKit/ASMapNode.mm

```diff
@@ -0,0 +1,210 @@
+      _options.size = self.calculatedSize;
+      _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+- (BOOL)isLiveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _liveMap;
+}
+
+- (void)setLiveMap:(BOOL)liveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (liveMap == _liveMap) {
+    return;
+  }
+  _liveMap = liveMap;
+  liveMap ? [self addLiveMap] : [self removeLiveMap];
+}
+
+
+- (BOOL)needsMapReloadOnBoundsChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _needsMapReloadOnBoundsChange;
+}
+
+- (void)setNeedsMapReloadOnBoundsChange:(BOOL)needsMapReloadOnBoundsChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  _needsMapReloadOnBoundsChange = needsMapReloadOnBoundsChange;
+}
+
+- (void)fetchData
+{
+  [super fetchData];
+  if (_liveMap && !_mapNode) {
+    [self addLiveMap];
+  }
+  else {
+    [self setUpSnapshotter];
+    [self takeSnapshot];
+  }
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  [self removeLiveMap];
+}
+
+- (void)takeSnapshot
+{
+  if (!_snapshotter.isLoading) {
+    [_snapshotter startWithCompletionHandler:^(MKMapSnapshot *snapshot, NSError *error) {
+      if (!error) {
+        UIImage *image = snapshot.image;
+        CGRect finalImageRect = CGRectMake(0, 0, image.size.width, image.size.height);
+        
+        UIGraphicsBeginImageContextWithOptions(image.size, YES, image.scale);
+        [image drawAtPoint:CGPointMake(0, 0)];
+        
+        if (_annotations.count > 0 ) {
+          // Get a standard annotation view pin. Future implementations should use a custom annotation image property.
+          MKAnnotationView *pin = [[MKPinAnnotationView alloc] initWithAnnotation:nil reuseIdentifier:@""];
+          UIImage *pinImage = pin.image;
+          for (id<MKAnnotation>annotation in _annotations)
+          {
+            CGPoint point = [snapshot pointForCoordinate:annotation.coordinate];
+            if (CGRectContainsPoint(finalImageRect, point))
+            {
+              CGPoint pinCenterOffset = pin.centerOffset;
+              point.x -= pin.bounds.size.width / 2.0;
+              point.y -= pin.bounds.size.height / 2.0;
+              point.x += pinCenterOffset.x;
+              point.y += pinCenterOffset.y;
+              [pinImage drawAtPoint:point];
+            }
+          }
+        }
+        
+        UIImage *finalImage = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        self.image = finalImage;
+      }
+    }];
+  }
+}
+
+- (void)resetSnapshotter
+{
+  if (!_snapshotter.isLoading) {
+    _options.size = self.calculatedSize;
+    _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+#pragma mark - Action
+- (void)addLiveMap
+{
+  if (self.isNodeLoaded && !_mapNode) {
```

Comment:
@appleguy 
- It doesn't work to set isLiveMap before the node is loaded since it's protected by the if condition checking if the node is loaded. I think to implement this behaviour we would have to implement -didLoad and check if isLiveMap = YES then call -addLiveMap. Having thought about this, I think we should support this functionality since otherwise there isn't any other way to start of with a live map. The problem is that since ASMapNode inherits from ASControlNode (indirectly) then it doesn't respond to touches until a target/action is added, so I need to set `self.userInteractionEnabled = YES;` directly to achieve this but I think this is acceptable.
- I've ran the current version of ASMapNode through Instruments and they seem to be getting destroyed properly, however I presume it would be better to have weakSelf anyway so I will do that.
- I've went with init with CGRectZero and calling setNeedsLayout since this feels cleaner.